crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -0,0 +1,1356 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(dead_code)]
+
+use std::{
+    fmt,
+    marker::PhantomData,
+    ops::Not,
+    ptr::NonNull,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+/// Errors of [`LinkedChunk`].
+#[derive(Debug)]
+pub enum LinkedChunkError {
+    InvalidChunkIdentifier { identifier: ChunkIdentifier },
+    ChunkIsAGap { identifier: ChunkIdentifier },
+    ChunkIsItems { identifier: ChunkIdentifier },
+    InvalidItemIndex { index: usize },
+}
+
+/// The [`LinkedChunk`] structure.
+///
+/// It is similar to a linked list, except that it contains many items instead
+/// of a single one. A chunk has a maximum capacity of `CAPACITY`. Once a chunk
+/// is full, a new chunk is created. Not all chunks are necessarily entirely
+/// full.
+pub struct LinkedChunk<T, const CHUNK_CAPACITY: usize> {
+    /// The first chunk.
+    first: NonNull<Chunk<T, CHUNK_CAPACITY>>,
+    /// The last chunk.
+    last: Option<NonNull<Chunk<T, CHUNK_CAPACITY>>>,
+    /// The number of items hold by this linked chunk.
+    length: usize,
+    /// The generator of chunk identifiers.
+    chunk_identifier_generator: ChunkIdentifierGenerator,
+    /// Marker.
+    marker: PhantomData<Box<Chunk<T, CHUNK_CAPACITY>>>,
```

Comment:
As far as I know, it's not mandatory, but it's always a good idea when manipulating `NonNull` to have a marker alongside with it. It has to do with some compiler optimisations or stuff like that. In case, I prefer to have it!